### DISCUSSION TITLE
Is it possible to manually get `SystemParam` to call non system functions?

### DISCUSSION THREAD

**Discussion by Bastczuak** [bevyengine/bevy#5482] [Q&A] [answered]
Hello,

is it possible in bevy_ecs to get a `SystemParam` manually from a world? I'm asking because I use opengl which is a `c_void` pointer and can't be used for bevys system because of missing `send + sync`. I give you an example.

```rs
main() {
  ...
  gameloop {
    ...
    schedule.run(&mut world);
    // call a function manually to render after world run
    render(&gl, world.get_system_param::<RenderData>()
  }
}

#[derive(SystemParam)]
struct RenderData<'w, 's> {
  camera: Res<'w, Camera>,
  transforms : Query<'w, 's, &'static Transforms>,
}

// not a registered system
pub fn render(opengl: &Gl, render_data: Renderdata) {...}
```

**Comment by alice-i-cecile** [comment, marked as answer]
Does [SystemState](https://dev-docs.bevyengine.org/bevy/ecs/system/struct.SystemState.html) do what you need?

**Reply by Bastczuak** [reply]
Yes that did the trick! Thanks!

```rs

type RenderSystemState<'w, 's> =(
  Res<'w, Camera>,
  Query<'w, 's, &'static Transform>,
);

main() {
  ...
  let mut render_state = SystemState::<RenderSystemState>::new(&mut world);
  ...
  gameloop {
    ..
    schedule.run(&mut world);
    render(&gl, render_state.get_mut(&mut world));
  }
}

pub fn render(opengl: &Gl, render_state: RenderSystemState) {
    let (camera, query) = render_state;
}
```
Im not sure doing the lifetimes correct for my type?! Can you please check? Can I provide an official example in this repo?

**Reply by alice-i-cecile** [reply]
That looks correct, but you should probably use the SystemParam derive macro for this :)

If we don't have an example for custom system param we should.

**Reply by Bastczuak** [reply]
Where should I use `SystemParam`? Instead of my own type? I'm lost again.

**Reply by alice-i-cecile** [reply]
```rust
#[derive(SystemParam)]
struct RenderingData<'w, 's> {
  camera: Res<'w, Camera>,
  transform_data: Query<'w, 's, &'static Transform>,
);
```

Like so :)

Although that's not strictly needed; you can just use

```rust
  let mut render_state = SystemState::<(Res<Camera>, Query<&Transform>)>::new(&mut world);
```
